src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs

```diff
@@ -12,13 +16,81 @@ internal static class WebInputSources
         /// </summary>
         public static ImmutableHashSet<SourceInfo> SourceInfos { get; }
 
+        /// <summary>
+        /// Cached information if the specified symbol is a Asp.Net Core Controller: (compilation) -> ((class symbol) -> (is Controller))
+        /// </summary>
+        private static readonly BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>> s_classIsControllerByCompilation =
+            new BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>>();
+
+        /// <summary>
+        /// Cached information if the specified symbol is a Asp.Net Core Action: (compilation) -> ((method symbol) -> (is Action))
+        /// </summary>
+        private static readonly BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>> s_methodIsActionByCompilation =
```

Comment:
> I have added a second cache to address full recompilation too. Let me know if you think it is overkill.

What did you mean by a second cache?

I'm still skeptical of caching by method. I'd prefer not to have a cache grow for each method in code unless that's absolutely necessary. I'd be okay if the methodCache only had entries for methods within a controller. Of course, it's hard to say for sure what's better without data from performance testing and trying different approaches.

I was thinking more along the lines of:
- If `parameter.ContainingSymbol` isn't an `IMethodSymbol`, or `parameter.ContainingSymbol.ContainingSymbol` isn't a `INamedTypeSymbol`, then return false.
- For the `INamedTypeSymbol` of the type, check the classCache if it's a controller or not.
- If not found in the cache:
  - Determine if the type is a controller, add to classCache.
- If type is not a controller, return false.
- If the method isn't action, return false.
- If the parameter has `[FromServices]`, return false.
- Return true.